%config IPCompleter.greedy=True


In [1]:
# square number upto a given number

new_list=[]

for x in range(1,11):
    new_list.append(x**2)

print (*new_list)

1 4 9 16 25 36 49 64 81 100


In [2]:
# prime number upto a given number

def prime_numbers_upto(a_number):
    for x in range(1, a_number):
        count =0
        if x <= 2:
            print(x, end=" ")
    
        elif(x%2 !=0):
            for y in range (2, x//2):
                if x%y ==0:
                    count +=1
            if count == 0:
                print (x, end =" ")
                

prime_numbers_upto(21)                
        

1 2 3 5 7 11 13 17 19 

In [3]:
# turning list to a string

a_list = ["a","b","c"]

a_str_1 = "".join(a_list)
a_str_2 = " ".join(a_list)
print(a_str_1)
print(a_str_2)

abc
a b c


In [4]:
# turning string to a list

a_str_1 = "abc"
a_str_2 = "a b c"
a_list_1 = list(a_str_1)
a_list_2 = list(a_str_2)

print(a_list_1)
print(a_list_2)

['a', 'b', 'c']
['a', ' ', 'b', ' ', 'c']


In [5]:
# side effects after mutation 

warm = ["yellow", "orange"]
hot = ["red"]

colors = [warm]
print(colors)

colors.append(hot)
print(colors)

hot.append("pink")
print(hot)
print(colors)

[['yellow', 'orange']]
[['yellow', 'orange'], ['red']]
['red', 'pink']
[['yellow', 'orange'], ['red', 'pink']]


In [6]:
# removing common elements of a list
# why returned [2,3,4] instead of [3,4]???
# be careful, do not mutate a list that you iterate over!!!

list_1 = [1,2,3,4]
list_2 = [1,2,5,6]

for n in list_1:
    if n in list_2:
        list_1.remove(n)
        
print(list_1)

[2, 3, 4]


In [7]:
# correction to above problem
# first get copy of the list, iterate over the copy and mutate the original list

list_1 = [1,2,3,4]
list_2 = [1,2,5,6]
list_1_copy = list_1 [:] # get copy of the list

for n in list_1_copy:
    if n in list_2:
        list_1.remove(n)
        
print(list_1)

[3, 4]


## Sets

The sets module provides classes for constructing and manipulating **unordered collections of unique elements**. We can use it to compare two lists.

In [8]:
# s = {} creates empty dictionary not a set
# s = ([]) creates an empty set

set_1 = set([1,2,3,4])
set_2 = {2,4,6,8}
set_3 = set([])
set_4 = {1,2,1,2,3,3,0,1}

print(set_1, set_2, set_3, set_4)

{1, 2, 3, 4} {8, 2, 4, 6} set() {0, 1, 2, 3}


In [9]:
# add an item to a set

set_1.add(5)
set_1

{1, 2, 3, 4, 5}

In [10]:
# add multiple items to a set

set_1.update([4,5,6,7,8,9])
set_1

{1, 2, 3, 4, 5, 6, 7, 8, 9}

In [11]:
# remove an item from a set
# we can use both "remove()" and "discard()" method to remove an item from a set.
# If the element is not a member, "remove()" raises a KeyError, whereas "discard()" doesn't. 

set_1.remove(9)
set_1.discard(8)
set_1.discard(10)
set_1

{1, 2, 3, 4, 5, 6, 7}

In [12]:
# finding intersection among sets

s1 = {1, 2, 3}
s2 = {2, 3, 4}
s3 = {3, 4, 5}

s4= s1.intersection(s2,s3)
s4

{3}

In [13]:
# finding differences among sets

s1 = {1, 2, 3}
s2 = {2, 3, 4}
s3 = {3, 4, 5}

s4 = s1.difference(s2)
s5 = s2.difference(s1)
s6 = s2.difference(s1, s3)

print(s4)
print(s5)
print(s6)

{1}
{4}
set()


In [14]:
# finding symmetric differences among sets
# set of elements that are in either A or B , but not in their intersection.

s1 = {1, 2, 3}
s2 = {2, 3, 4}
s3 = {3, 4, 5}

s4 = s1.symmetric_difference(s2)

print(s4)

{1, 4}


## Comprehensions

In [15]:
# multiply each item in the list by 10 and make another list

nums = [1, 2, 3, 4, 5]
my_list = [n*10 for n in nums]
my_list

[10, 20, 30, 40, 50]

In [16]:
# multiply each item in the list by 10 if its an odd number 
# and make another list

nums = [1, 2, 3, 4, 5]
my_list = [n*10 for n in nums if n%2==1]
my_list

[10, 30, 50]

In [17]:
# odd numbers upto 10

x = [x for x in range(10) if x%2==1]
print (x)

[1, 3, 5, 7, 9]


In [18]:
# (letter, num) pair for each letter in "x,y,z" and each number in "0,1,2,3"

my_list = [(letter, num) for letter in "xyz" for num in range(4)]
my_list

[('x', 0),
 ('x', 1),
 ('x', 2),
 ('x', 3),
 ('y', 0),
 ('y', 1),
 ('y', 2),
 ('y', 3),
 ('z', 0),
 ('z', 1),
 ('z', 2),
 ('z', 3)]

In [19]:
# seperate words into letters

def letter_seperator(a_str):
    x = [a_str[n] for n in range(len(a_str))]
    print(x)
    
letter_seperator("I know how to seperate words!")

['I', ' ', 'k', 'n', 'o', 'w', ' ', 'h', 'o', 'w', ' ', 't', 'o', ' ', 's', 'e', 'p', 'e', 'r', 'a', 't', 'e', ' ', 'w', 'o', 'r', 'd', 's', '!']


In [20]:
# nested lists

nested_list=[[1,2,3,4,5],[6,7,8,9,10]]

x = [item for inner_list in nested_list for item in inner_list]
print(x)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


In [21]:
# dictionary comprehensions

{x:x**2 for x in range(1,10) if x%2==0}

{2: 4, 4: 16, 6: 36, 8: 64}

In [22]:
# dictionary comprehensions

first = ['Ali','Veli','Cem', "David", "Elif", "Fahri"]
last = ['Bir','Yedi','Sekiz', "Yedi", "Bir", "Sekiz"]

my_dict = {first:last for first, last in zip(first, last)}
my_dict

{'Ali': 'Bir',
 'Veli': 'Yedi',
 'Cem': 'Sekiz',
 'David': 'Yedi',
 'Elif': 'Bir',
 'Fahri': 'Sekiz'}

In [23]:
# dictionary comprehensions

first = ['Ali','Veli','Cem', "David", "Elif", "Fahri"]
last = ['Bir','Yedi','Sekiz', "Yedi", "Bir", "Sekiz"]

my_dict = {first:last for first, last in zip(first, last) if last != "Yedi"}
my_dict

{'Ali': 'Bir', 'Cem': 'Sekiz', 'Elif': 'Bir', 'Fahri': 'Sekiz'}

In [24]:
# set comprehensions

nums = [1,2,3,1,2,3,5,6,4,5,6,4,8,9,7,8,9,7,8]
my_set= set(nums)
my_set

{1, 2, 3, 4, 5, 6, 7, 8, 9}

In [25]:
# set comprehensions

nums = [1,2,3,1,2,3,5,6,4,5,6,4,8,9,7,8,9,7,8]
my_set= {n for n in nums}
my_set

{1, 2, 3, 4, 5, 6, 7, 8, 9}

### lambda
lambda functions are another way of defining the functions
it can be useful when you need to create a function that is going to be used only once 
or when you don't want to bother defining all the routines for function

In [26]:
sqr = lambda x:x**2
sqr(4)

16

In [27]:
is_result_even = lambda x,y : True if x*y%2==0 else False

print(is_result_even(3,5))
is_result_even(4,7)

False


True

### filter

The filter filters out items based on a test function which is a filter and apply functions to pairs of item 
The first parameter in filter () must be boolean
This function will be applied to every element of the list l. 
Only if first parameter returns True will the element of the list be included in the result list.

In [28]:
def is_even(x):
    
    if x%2==0:
        
        return True
    else:
        return False


list(filter (is_even, range(1,11)))
# there is no () or parameter when calling the function

[2, 4, 6, 8, 10]

In [29]:
# the same with lambda fıunction

list(filter (lambda x: True if x % 2==0 else False, range(1,11)))

[2, 4, 6, 8, 10]

In [30]:
# the same with lambda fıunction - shorter version
list(filter (lambda x: x % 2==0, range(1,11)))

[2, 4, 6, 8, 10]

### map
map function applies a passed-in function to each item in an iterable object 
and returns a list containing all the function call result.
it takes a function as the first parameter, and an iterable object as the second parameter

In [31]:
items = [1, 2, 3, 4, 5]

def sqr(x): 
    
    return x ** 2

list(map(sqr, items))

[1, 4, 9, 16, 25]

In [32]:
list(map(lambda x: x**2, items))

[1, 4, 9, 16, 25]

In [33]:
# the squares if the number is even; else leave it as it is

result = list(map((lambda x: x**2 if x%2==0 else x),  range(10)))
print (*result)

0 1 4 3 16 5 36 7 64 9


## random module

In [34]:
# random() method
# returns a value 0(inclusive) and 1 (not inclusive) 

import random

value = random.random()
value

0.9656339558399094

In [35]:
# uniform method()
# returns a floating between a specified range

value = random.uniform(4,13)
value

9.043275939554588

In [36]:
# randint method()
# returns an integer between a specified range

value = random.randint(-4,13)
value

9

In [37]:
# choice() method
# picks a random value from a list of values

students_list = ['Ali','Veli','Cem', "David", "Elif", "Fahri"]

selected_student = random.choice(students_list)
print("Selected student is {}".format(selected_student))

Selected student is Ali


In [38]:
# choices() method
# It returns a list with the randomly selected element from the specified sequence.
# You can weigh the possibility of each result with the weights parameter.
# The sequence can be a string, a range, a list, a tuple or any other kind of sequence.

colors = ['red','blue','green']

results =  random.choices(colors, weights=[4,4,2], k=10)
print(results)

['red', 'blue', 'red', 'blue', 'red', 'red', 'red', 'green', 'green', 'red']


In [39]:
# be careful! even length of list is greater than the "k" value, choices() method may return the same values.
# for selecting unique values we use sample() method.
# When using sample() method, "k" parameter should be less than length of list

deck = list(range(1,53))

samples =  random.sample(deck, k=7)
print(samples)

[46, 25, 17, 51, 11, 18, 7]


In [40]:
# shuffle() method,  shuffles a list in place

deck = list(range(1,53))

random.shuffle(deck)
print(deck)

[23, 22, 2, 43, 52, 28, 3, 44, 32, 41, 10, 15, 49, 39, 9, 4, 24, 16, 46, 1, 48, 19, 36, 20, 50, 38, 13, 26, 29, 14, 18, 45, 31, 33, 21, 37, 40, 34, 42, 17, 27, 12, 25, 5, 30, 35, 6, 11, 51, 7, 47, 8]


## "==" vs "is" for comparison

* "==" checks for equality
* "is" checks for identity, if two objects hold the same location in memory  

In [41]:
list_1 = [1,2,3]
list_2 = [1,2,3]

print(id(list_1))
print(id(list_2))

print(list_1 == list_2)
print(list_1 is list_2)

2611413973128
2611413987528
True
False


In [42]:
list_1 = [1,2,3]
list_2 = list_1

print(id(list_1))
print(id(list_2))

print(list_1 == list_2)
print(list_1 is list_2)

2611413555464
2611413555464
True
True


### unpacking

In [43]:
# assign two variables to the items of a tuple at the same time

a, b = (1,2)
print(a,b)

1 2


In [44]:
# if you don't need the second variable

a, _ = (3,4)
print(a)

3


In [45]:
# assign multiple variables 

a, b, *c = (1,2,3,4,5,6,7)
print(a,b, c)

1 2 [3, 4, 5, 6, 7]


In [46]:
# assign multiple variables 

a, b, *c, d, e = (1,2,3,4,5,6,7)
print(a,b,c,d,e)

1 2 [3, 4, 5] 6 7


## zipfile module

In [47]:
import zipfile

my_zip = zipfile.ZipFile("zipped_files.zip", 'w')

my_zip.write("one.txt")
my_zip.write("two.png")

my_zip.close()

FileNotFoundError: [WinError 2] Sistem belirtilen dosyayı bulamıyor: 'one.txt'

In [ ]:
# Alternatively with context manager

with zipfile.ZipFile("zipped_files.zip", 'w') as my_zip:
    my_zip.write("one.txt")
    my_zip.write("two.png")

## 3 different methods for iteration

We will try to find a cube root of a given number with 3 different methods.

### 1. guess-and-check algorithms

* searches from 0 to the given number.
* when exceeds the the given number break the loop
* just give the answer if the given number is a perfect cube, otherwise returns nothing.

In [1]:
def find_cuberoot(cube):

    for num in range(abs(cube)+1):
        
        if num**3 >= abs(cube):
            break;

    if num**3 != abs(cube):
        print(f"{cube} isn't a perfect cube")
    
    else:
        if cube < 0:
            num = -num
        print (f"cube root of {cube} is {num}")

find_cuberoot(-27)

cube root of -27 is -3


### 2. approximate solution methods

* searching for a close enough answer
* starts from 0 and increment by some small value
* "close enough" is given by an "epsilon" number which defines solution interval
* keep iteration if `abs(num**3 - cube) >= epsilon` 
* decreasing increment size = > slower program
* increasing epsilon = > less acurate answer

In [2]:
def find_cuberoot(cube, increment_size, epsilon):
    
    num = 0
    no_trial = 0
    while abs(num**3 - cube) >= epsilon and num**3 <= cube:
        
        num += increment_size
        no_trial +=1
    print (f"have done {no_trial} trials ")
    
    if abs(num**3 > cube):
        print("couldn't find any solution")
    else:
        print(f"cube root of {cube} is approximately {num}")

find_cuberoot(100, 0.0001, 0.01)

have done 46415 trials 
cube root of 100 is approximately 4.641500000002521


### bisection search

* the list must be sorted
* eliminates half the search space with each iteration
* the larger the space is, the better it is to use bisection seacrh 

In [3]:
def find_cuberoot(cube, epsilon):
    
    no_trial = 0
    low =0
    high = cube
    num = (low+high)/2
    while abs(num**3 - cube) >= epsilon:
        
        if num**3 > cube:
            high=num
        else:
            low= num

        no_trial +=1
        num = (low+high)/2        
    
    print (f"have done {no_trial} trials ")
    
    print(f"cube root of {cube} is approximately {num}")

find_cuberoot(100, 0.01)

have done 14 trials 
cube root of 100 is approximately 4.6417236328125


* with approximate solution methods we have made 46415 trials to find cube root of 100, whereas we have made just 14 trials with bisection search method.

In [28]:
# bisection search with recursion method_ Option 1
# search whether an element exists in an ordered list
# complexity O(n log n), since we copy list in recursion methods

def bisect_search_1(a_list, e):
    
    # check if there is an element in the list
    if a_list==[]:
        return False
    
    # if there is only 1 element
    elif len(a_list)==1:
        return e == a_list[0]
    
    else: 
        
        # find the half point of the list 
        half = len(a_list) // 2
        
        # if the middle element is greater than what we search
        # search for first half, otherwise the second
        if a_list[half] > e:
            return bisect_search_1(a_list[:half], e)
        else:
            return bisect_search_1(a_list[half:], e)

bisect_search_1([1,2,3,4,5,6,7,8,9], 9)

True

In [32]:
# bisection search with recursion method_ Option 2
# search whether an element exists in an ordered list
# complexity O(log n), since we don't copy list in recursion methods, just move the list indexes

def bisect_search_2(a_list, e):
    def bisect_search_helper(a_list, e, low, high):
        
        # check if there is only 1 element left
        if high==low:
            return a_list[low] ==e
        
        # find the middle point of the list 
        middle = (low+high)//2
        
        if a_list[middle] == e:
            return True
        
        # if the middle element is greater than what we search
        # search for first half, otherwise the second
        elif a_list[middle] > e:
            if low == middle: # nothing left to search
                return False
            else: 
                return bisect_search_helper(a_list, e, low, middle-1)
        else:
            return bisect_search_helper(a_list, e, middle+1, high)
    
    # check if there is an element in the list
    if len(a_list) == 0:
        return False
    else:
        return bisect_search_helper(a_list, e, 0, len(a_list))
    
bisect_search_2([1,2,3,4,5,6,7,8,9], 9)

True

## Recursion

* reduce a problem to simpler versions of the same problem
* a programming technique where a function calls itself with a reduced input

In [4]:
# "a" multiplied by "b"

def multiplication(a, b):
    
    # base case
    if b ==1:
        return a
    
    # recursive step
    else: 
        return a+ multiplication(a, b-1)
    
multiplication(5,10)

50

In [5]:
# calculate n factorial with recursion method

def factorial(n):
    
    # base case
    if n ==1:
        return 1
    
    # recursive step
    else: 
        return n * factorial(n-1)
    
factorial(10)

3628800

In [6]:
# calculate n factorial with iteration method

def factorial(n):
    
    result=1
    for n in range(1, n+1):
        result *= n
    
    return result
        
factorial(10)

3628800

In [16]:
# fibonacci numbers with recursive method
# 1-1-2-3-5-8-13-21

import time

def fib(n):
    if n==0 or n==1:
        return 1
    else:
        return fib(n-1) + fib(n-2)

t_0 = time.process_time()

print(fib(34))

t_1 = time.process_time() - t_0

print(f"time elapsed : {t_1}")

9227465
time elapsed : 2.234375


In [14]:
# fibonacci with a dictionary
# first initialize the dictionary then call function
# with the above method, to calculate fib(34) it makes over 11 million recurcive calss,
# with using a dictionary, it makes just 65 recurcive calss,

def fib(n, a_dict):
    if n in a_dict:
        return a_dict[n]
    else:
        result = fib(n-1, a_dict) + fib(n-2, a_dict)
        a_dict[n] = result
        return result


t_0 = time.process_time()

fib_dict = {1:1, 2:2}
print(fib(34, fib_dict))

t_1 = time.process_time() - t_0

print(f"time elapsed : {t_1}")

9227465
time elapsed : 0.0


In [1]:
# palindrome
# sequence of characters which reads the same backward as forward

def is_Palindrome(a_str):

    def to_Chars(a_str):
        a_str = a_str.lower()
        result =""
        for letter in a_str:
            if letter in "abcdefghijklmnopqrstuvwxyz":
                result += letter
        return result
        
    def is_Pal(a_str):
        if len(a_str) <=1:
            return True
        else:
            return a_str[0] == a_str[-1] and is_Pal(a_str[1:-1])
    
    return is_Pal(to_Chars(a_str))
        
            
print(is_Palindrome("A man, a plan, a canal – Panama"))
print(is_Palindrome("Able was I ere I saw Elba"))

True
True


In [28]:
# POWER SET – CONCEPT 
# to generate the power set	of integers from 1 to n
# assume we can generate power set of integers from 1 to n-1	
# then	all	of	those subsets belong to	bigger	power set and all of those subsets with	n	
# added to each	of them also belong	to the bigger power set	


def genSubsets(a_list):
    
    if len(a_list) == 0:
        return [[]] # list of empty list
    
    # all subsets without last element
    smaller = genSubsets(a_list[:-1]) 
    
    # last element as a list 
    extra = [a_list[-1]]
    new =[]
    
    # last element will be added to all subsets
    for small in smaller:
        new.append(small+extra)
    
    # combine those with last element and those without
    return smaller + new

genSubsets([1,2,3,4])

[[],
 [1],
 [2],
 [1, 2],
 [3],
 [1, 3],
 [2, 3],
 [1, 2, 3],
 [4],
 [1, 4],
 [2, 4],
 [1, 2, 4],
 [3, 4],
 [1, 3, 4],
 [2, 3, 4],
 [1, 2, 3, 4]]

## Sorting

#### Selection Sort

1. extract minimum element, swap it with element at index 0
2. in remaining sublist, extract minimum element, swap it with the element	at index 1		
3. keep	the	left portion of	the	list sorted, at	i’th step, first i	elements in list are sorted, all other elements are	bigger than first i elements

complexity:  O($n^2$) where n is len(L)	

In [41]:
def selection_sort(L):
    suffixSt = 0
    while suffixSt != len(L):
        for i in range(suffixSt+1, len(L)):
            if L[i] < L[suffixSt]:
                L[suffixSt], L[i] = L[i], L[suffixSt]
        suffixSt += 1
    return L
selection_sort([4,3,0,2,1,6,9,8,7,5,5,0])

[0, 0, 1, 2, 3, 4, 5, 5, 6, 7, 8, 9]

#### Merge Sort

	use a divide-and-conquer approach:
1. if list is of length	0 or 1,	already	sorted	
2. if list has more	than one element, split	into two lists, and sort each	
3. merge	sorted	sublists	
    a. look	at	first	element	of	each,	move	smaller	to	end	of	the result	
    b. when	one	list	empty,	just	copy	rest	of	other	list	
    
complexity:  O(n log n) where n is len(L)	

In [43]:
def merge(left, right):
    result = []
    i,j = 0,0
    while i < len(left) and j < len(right):
        if left[i] < right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1
    while (i < len(left)):
        result.append(left[i])
        i += 1
    while (j < len(right)):
        result.append(right[j])
        j += 1
    return result


def merge_sort(L):
    if len(L) < 2:
        return L[:]
    else:
        middle = len(L)//2
        left = merge_sort(L[:middle])
        right = merge_sort(L[middle:])
    return merge(left, right)
merge_sort([4,3,0,2,1,6,9,8,7,5,5,0])

[0, 0, 1, 2, 3, 4, 5, 5, 6, 7, 8, 9]